### PR TITLE
feat(tsql): Add support for scope qualifier operator

### PR DESCRIPTION
**Pull request by VaggelisD**
Fixes [#SQLMesh/2817](https://github.com/TobikoData/sqlmesh/issues/2817)

Introduce support for `<data type>::<static_func>` in T-SQL.

### CODE REVIEW

**Review by georgesittas — COMMENTED** [review]
This doesn't look quite right, we shouldn't handle just geography.

From what I've seen T-SQL doesn't use the `::` syntax for casting, so imo we should just create a new expression like `ScopeQualifier` to represent it and override the corresponding `COLUMN_OPERATORS` entry so that it produces it. That would yield a much simpler solution AFAICT.

See also: https://www.sqlteam.com/forums/topic.asp?TOPIC_ID=76723

**Comment by georgesittas** [comment]
So from what I'm seeing `ScopeResolution` would be a binary in this case and it should just work

**Comment by VaggelisD** [comment]
Sounds good, not sure why I remembered that `::` was a cast operator in T-SQL but it doesn't seem to be. 

Besides that, I'm not sure if the solution can be simplified much further because in `parser.py::_parse_column_ops()` the `TokenType.DCOLON` operator is hardcoded to parse types, so in `tsql.py::COLUMN_OPERATORS` the RHS expression will be a "user-defined" `exp.DataType` with the function params as `exp.DataTypeParams` instead of a single `exp.Func`, right?

**Comment by georgesittas** [comment]
Yeah, I see what you mean. Maybe we should challenge the idea of keeping this as a "column operator" while we're at it. For instance, I saw the `::foo` syntax in one of the links I shared. If we can pop it from `COLUMN_OPERATORS` and handle it elsewhere then we won't have the issue you mentioned.

**Comment by VaggelisD** [comment]
I saw the `::foo` syntax on the last reply of the 1st link (forum post) but in the same reply the author claims it's needed anymore after SQL Server 2005, not sure if you're referring to the same one.

If that's the case, then maybe we can preserve the "binary column operator" solution, although not semantically correct? I'm not seeing other ways out of it if we want to redesign it from scratch, will have to search this a bit better

**Comment by georgesittas** [comment]
Can't we just use a range parser?

**Comment by VaggelisD** [comment]
`RANGE_PARSER` works yeah, probably not semantically correct as well but it did simplify things.

File: sqlglot/expressions.py [hunk 1]
```diff
@@ -5580,6 +5580,11 @@ class MapFromEntries(Func):
     pass
 
 
+# https://learn.microsoft.com/en-us/sql/t-sql/language-elements/scope-resolution-operator-transact-sql?view=sql-server-ver16
+class ScopeResolution(Func):
```
Comment (georgesittas) [inline sqlglot/expressions.py:5584]: `Func` doesn't look right here - let's opt for `Expression` instead.

File: sqlglot/dialects/tsql.py [hunk 2]
```diff
@@ -524,6 +524,13 @@ class Parser(parser.Parser):
             "TIMEFROMPARTS": _build_timefromparts,
         }
 
+        RANGE_PARSERS = {
+            **parser.Parser.RANGE_PARSERS,
+            TokenType.DCOLON: lambda self, this: self.expression(
+                exp.ScopeResolution, this=this, expression=self._parse_function()
```
Comment (georgesittas) [inline sqlglot/dialects/tsql.py:530]: Let's use a more lenient parser for the RHS, eg. the 2nd [link](https://learn.microsoft.com/en-us/sql/t-sql/statements/grant-server-principal-permissions-transact-sql?view=sql-server-ver16) shows this syntax:

```sql
GRANT VIEW DEFINITION ON LOGIN::EricKurjan TO RMeyyappan   
    WITH GRANT OPTION
```

**Review by georgesittas — APPROVED** [review]
LGTM, couple of comments and should be good to go.

> probably not semantically correct as well but it did simplify things

Why not?

File: tests/dialects/test_tsql.py [hunk 3]
```diff
@@ -1878,3 +1878,11 @@ def test_declare(self):
             "DECLARE vendor_cursor CURSOR FOR SELECT VendorID, Name FROM Purchasing.Vendor WHERE PreferredVendorStatus = 1 ORDER BY VendorID",
             check_command_warning=True,
         )
+
+    def test_scope_resolution_op(self):
+        self.validate_identity("SELECT ::FOO(a, b)")
+        self.validate_identity("SELECT bar::BAZ(1, 2)")
+        self.validate_identity("SELECT GEOGRAPHY::POINT([latitude], [longitude], 4326) FROM tbl")
```
Comment (georgesittas) [inline tests/dialects/test_tsql.py:1885]: This is problematic after all, T-SQL is case-sensitive when it comes to these system functions:

```
1> SELECT GEOGRAPHY::POINT(34.2, 34.2, 4326)
2> go
Msg 6506, Level 16, State 10, Server af4359a4e415, Line 1
Could not find method 'POINT' for type 'Microsoft.SqlServer.Types.SqlGeography' in assembly 'Microsoft.SqlServer.Types'
1> SELECT GEOGRAPHY::Point(34.2, 34.2, 4326)
2> go
  
--
0x
```

Comment (georgesittas) [reply tests/dialects/test_tsql.py:1885]: Maybe we can switch off the function name normalization only for anonymous functions in T-SQL and control this behavior with a generator flag?

Comment (georgesittas) [reply tests/dialects/test_tsql.py:1885]: Or if we can prove this is the only case where case-sensitivity matters, check the parent of the `Func` and only normalize it if it's not a `ScopeResolution`.

Comment (VaggelisD) [reply tests/dialects/test_tsql.py:1885]: I tried a few different functions like date/time related ones and they all seem to be case insensitive except these, will disable normaliaztion for `ScopeResolution` funcs

Comment (georgesittas) [reply tests/dialects/test_tsql.py:1885]: Cool, let's only do this for T-SQL

Comment (VaggelisD) [reply tests/dialects/test_tsql.py:1886]: It's awkward to _not_ normalize some of the functions in a dialect; For example, as I mentioned in T-SQL the builtin functions like `DATEDIFF`, `CONVERT` etc are case-insensitive in contrast to the CLR functions like `geography::Point` which are case-sensitive.

In order not to be intrusive to the rest of T-SQL, I chose to manually generate the function RHS of `exp.ScopeResolution`; If we want to normalize all functions in T-SQL I can follow up with a commit and change the PR to breaking.

**Review by georgesittas — APPROVED** [review]

**Pull request merged**